File: springboot/src/main/java/br/com/biblioteca/dto/ApiResponse.java
package br.com.biblioteca.dto;

public class ApiResponse<T> {
    private boolean sucesso;
    private String mensagem;
    private T dado;

    public ApiResponse(boolean sucesso, String mensagem, T dado) {
        this.sucesso = sucesso;
        this.mensagem = mensagem;
        this.dado = dado;
    }

    // Fábrica de sucesso
    public static <T> ApiResponse<T> sucesso(T dado) {
        return new ApiResponse<>(true, "Operação realizada com sucesso", dado);
    }

    public static <T> ApiResponse<T> sucesso(T dado, String mensagem) {
        return new ApiResponse<>(true, mensagem, dado);
    }

    // Fábrica de erro
    public static <T> ApiResponse<T> erro() {
        return new ApiResponse<>(false, "Ocorreu um erro", null);
    }

    public static <T> ApiResponse<T> erro(String mensagem) {
        return new ApiResponse<>(false, mensagem, null);
    }

    // Getters e Setters
    public boolean isSucesso() {
        return sucesso;
    }

    public void setSucesso(boolean sucesso) {
        this.sucesso = sucesso;
    }

    public String getMensagem() {
        return mensagem;
    }

    public void setMensagem(String mensagem) {
        this.mensagem = mensagem;
    }

    public T getDado() {
        return dado;
    }

    public void setDado(T dado) {
        this.dado = dado;
    }
}

File: angular/src/app/shared/dto/api-response.ts
export class ApiResponse<T> {
  sucesso: boolean;
  mensagem: string;
  dado: T | null;

  constructor(sucesso: boolean, mensagem: string, dado: T | null) {
    this.dado = dado;
    this.mensagem = mensagem;
    this.sucesso = sucesso;
  }

  static sucesso<T>(dado: T, mensagem = 'Operação realizada com sucesso'): ApiResponse<T> {
    return new ApiResponse<T>(true, mensagem, dado);
  }

  static erro<T>(mensagem = 'Ocorreu um erro'): ApiResponse<T> {
    return new ApiResponse<T>(false, mensagem, null);
  }

}
